// @serenity-js/core/lib/io/file_system.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("graceful-fs");
var mkdirp = require("mkdirp");
var path = require("path");
var FileSystem = (function () {
    function FileSystem(root) {
        this.root = root;
    }
    FileSystem.prototype.store = function (relativePathToFile, data) {
        var _this = this;
        return this.ensureSpecified(relativePathToFile)
            .then(function (relativePath) { return _this.prepareDirectory(relativePath); })
            .then(function (absolutePath) { return _this.write(absolutePath, data); });
    };
    FileSystem.prototype.ensureSpecified = function (relativePathToFile) {
        return !!(relativePathToFile && relativePathToFile.length)
            ? Promise.resolve(relativePathToFile)
            : Promise.reject(new Error('Please specify where the file should be saved'));
    };
    FileSystem.prototype.prepareDirectory = function (relativePathToFile) {
        var absolutePath = path.resolve(this.root, relativePathToFile), parent = path.dirname(absolutePath);
        return new Promise(function (resolve, reject) {
            mkdirp(parent, function (error) {
                if (error) {
                    reject(error);
                }
                resolve(absolutePath);
            });
        });
    };
    FileSystem.prototype.write = function (path, data) {
        return new Promise(function (resolve, reject) {
            fs.writeFile(path, data, function (error) {
                if (error) {
                    reject(error);
                }
                resolve(path);
            });
        });
    };
    return FileSystem;
}());
exports.FileSystem = FileSystem;
